### PR TITLE
Orion/create release 6

### PR DESCRIPTION
**Pull request by borland**
feat: Handle unresolved package versions and dynamic packages
feat: remove --package-prerelease
refactor: Use the common flag.GenerateAutomationCmd facility

### CODE REVIEW

File: pkg/cmd/release/create/create.go [hunk 1]
```diff
@@ -74,88 +110,67 @@ func NewCmdCreate(f factory.Factory) *cobra.Command {
...
+	flagAliases := make(map[string][]string, 10)
+	util.AddFlagAliasesString(flags, FlagGitRef, flagAliases, FlagAliasGitRefRef, FlagAliasGitRefLegacy)
+	util.AddFlagAliasesString(flags, FlagGitCommit, flagAliases, FlagAliasGitCommitLegacy)
+	util.AddFlagAliasesString(flags, FlagPackageVersion, flagAliases, FlagAliasDefaultPackageVersion, FlagAliasPackageVersionLegacy, FlagAliasDefaultPackageVersionLegacy)
+	util.AddFlagAliasesString(flags, FlagReleaseNotes, flagAliases, FlagAliasReleaseNotesLegacy)
+	util.AddFlagAliasesString(flags, FlagVersion, flagAliases, FlagAliasReleaseNumberLegacy)
+	util.AddFlagAliasesBool(flags, FlagIgnoreExisting, flagAliases, FlagAliasIgnoreExistingLegacy)
+	util.AddFlagAliasesBool(flags, FlagIgnoreChannelRules, flagAliases, FlagAliasIgnoreChannelRulesLegacy)
+
+	cmd.PreRunE = func(cmd *cobra.Command, args []string) error {
```
Comment (borland) [inline pkg/cmd/release/create/create.go:142]: I worked out a much nicer way of dealing with flag aliases. Could probably tighten it up a bit further, but I think this should do for now.

Logic might not be particularly clear so let me know if you'd like a run-through @domenicsim1 @jbristowe

File: pkg/cmd/release/create/create.go [hunk 2]
```diff
@@ -276,6 +282,19 @@ func BuildPackageVersionBaseline(octopus *octopusApiClient.Client, deploymentPro
...
+
+		// If a package is not considered resolvable by the server, don't attempt to query it's feed or lookup
+		// any potential versions for it; we can't succeed in that because variable templates won't get expanded
+		// until deployment time
+		if !pkg.IsResolvable {
+			result = append(result, &StepPackageVersion{
+				PackageID:            pkg.PackageID,
+				ActionName:           pkg.ActionName,
+				PackageReferenceName: pkg.PackageReferenceName,
+				Version:              "",
```
Comment (borland) [inline pkg/cmd/release/create/create.go:294]: This is the "magic" needed to deal with dynamic packages. Worked out nicely

File: pkg/cmd/release/create/create.go [hunk 3]
```diff
@@ -342,16 +361,14 @@ func BuildPackageVersionBaseline(octopus *octopusApiClient.Client, deploymentPro
...
-					// if channel rules have altered our query, tell the user about that
-					channelRulesHelp := ""
-					if query.PreReleaseTag != "" {
-						channelRulesHelp = fmt.Sprintf("%s, pre-release tag matching %s", channelRulesHelp, query.PreReleaseTag)
-					}
-					if query.VersionRange != "" {
-						channelRulesHelp = fmt.Sprintf("%s, version range matching %s", channelRulesHelp, query.VersionRange)
-					}
-					return nil, fmt.Errorf("no package version found for %s%s. please check that the package exists in your package feed", packageRef.PackageID, channelRulesHelp)
+				case 0: // no package found; cache the response
```
Comment (borland) [inline pkg/cmd/release/create/create.go:364]: Now that we consider "no available versions for this package" to be a legit thing, we need to cache that too

File: pkg/cmd/release/create/create.go [hunk 4]
```diff
@@ -382,15 +399,20 @@ type PackageVersionOverride struct {
 // This is the inverse of ParsePackageOverrideString
 func (p *PackageVersionOverride) ToPackageOverrideString() string {
 	components := make([]string, 0, 3)
-	if p.PackageReferenceName != "" {
-		components = append(components, p.PackageReferenceName)
-	}
+
+	// stepNameOrPackageID always comes first if we have it
```
Comment (borland) [inline pkg/cmd/release/create/create.go:403]: I had a bug here; I'd mis-interpreted the C# code in the server as being `packageReference:stepName:version` but it was actually `stepName:packageReferenceName:version`. See unit test changes etc

File: pkg/cmd/release/create/create.go [hunk 5]
```diff
@@ -658,9 +680,13 @@ func printPackageVersions(ioWriter io.Writer, packages []*StepPackageVersion) er
 	t.AddRow(output.Dim("PACKAGE"), output.Dim("VERSION"), output.Dim("STEPS"))
 
 	for _, pkg := range consolidated {
+		version := pkg.Version
+		if version == "" {
+			version = output.Red("ERROR") // can't determine version for this package
```
Comment (borland) [inline pkg/cmd/release/create/create.go:685]: It looks like this

image

File: pkg/cmd/release/create/create_test.go [hunk 6]
```diff
@@ -1066,7 +1202,7 @@ func TestReleaseCreate_AutomationMode(t *testing.T) {
 				ReleaseNotes:          "Here are some **release notes**.",
 				IgnoreIfAlreadyExists: true,
 				IgnoreChannelRules:    true,
```
Comment (borland) [inline pkg/cmd/release/create/create_test.go:1204]: As discussed, I've left IgnoreChannelRules, but removed PackagePrerelease

**Review by domenicsim1 — APPROVED** [review]

**Pull request merged**